mars/oscar/backends/mars/ray.py

```diff
@@ -0,0 +1,275 @@
+    # so we use the posixpath.
+    parts = []
+    if parsed_url.netloc:
+        tmp = parsed_url.path
+        while tmp and tmp != "/":
+            tmp, item = posixpath.split(tmp)
+            parts.append(item)
+    parts = list(reversed(parts))
+    if parts and len(parts) != 2:
+        raise ValueError(f"Only bundle index and process index path are allowed in ray "
+                         f"address {address}.")
+    name, bundle_index, process_index = [parsed_url.netloc] + parts if parts else ""
+    if bool(name) != bool(bundle_index) or bool(bundle_index) != bool(process_index):
+        raise ValueError(f"Missing placement group name or bundle index or process index "
+                         f"from address {address}")
+    if name and bundle_index:
+        return name, int(bundle_index), int(process_index)
+    else:
+        return name, -1, -1
+
+
+def pg_bundle_to_address(pg_name: str, bundle_index: int, process_index: int = 0):
+    return f"ray://{pg_name}/{bundle_index}/{process_index}"
+
+
+class NodeResourceSpec:
+    __slots__ = 'n_process', 'num_cpus', 'memory', 'num_gpus'
+
+    def __init__(self, n_process, num_cpus=None, num_gpus=None, memory=None):
+        self.n_process = n_process
+        self.num_cpus = num_cpus or n_process
+        assert n_process <= self.num_cpus
+        self.num_gpus = num_gpus
+        self.memory = memory
+
+    def to_bundle(self):
+        bundle = {'CPU': self.num_cpus}
+        if self.num_gpus:
+            bundle['GPU'] = self.num_gpus
+        if self.memory:
+            bundle['memory'] = self.memory
+        return bundle
+
+
+class ClusterManager:
+
+    def __init__(self, cluster_name, resource_specs: List[NodeResourceSpec]):
+        self.cluster_name = cluster_name
+        self.resource_specs = resource_specs
+        self.nodes_info = []
+        # Hold actor handle to avoid being freed.
+        self.main_pool_handles = [None] * len(resource_specs)
+
+    def add_node(self, index: int, main_handle, address):
+        pg_name, bundle_index, _process_index = address_to_placement_info(address)
+        assert len(self.nodes_info) == index
+        node_info = [pg_bundle_to_address(pg_name, bundle_index, process_index=i)
+                     for i in range(self.resource_specs[index].n_process + 1)]
+        self.nodes_info.append(node_info)
+        self.main_pool_handles[index] = main_handle
+
+    def get_node_info(self, index):
+        return self.nodes_info[index]
+
+    def get_all_nodes_info(self):
+        return list(self.nodes_info)
+
+    def addresses(self):
+        return [process for node in self.get_all_nodes_info() for process in node]
+
+    def address_to_resources(self) -> Dict[str, Dict[str, Number]]:
+        # TODO(chaokunyang) support address resources mapping
+        return {process: {"CPU": 1} for node in self.get_all_nodes_info() for process in node}
+
+
+def create_cluster(cluster_name, resource_specs: List[NodeResourceSpec]):
```

Comment:
Why defining a `create_cluster` here? There is a `setup_cluster` in `backends/ray/driver.py`.

Reply:
It's an convenience API for `deploy on ray`. `Mars actor pool on ray` use `MarsActorDriver. The whole implementation is based on `mars actor pool`. It's not very realted to implementation under `backends/ray`

Reply:
removed